src/Picocrypt.go

```diff
@@ -2330,9 +2365,105 @@ func sizeify(size int64) string {
 	}
 }
 
+func unpackArchive(zipPath string) error {
+	reader, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	var totalSize int64
+	for _, f := range reader.File {
+		totalSize += int64(f.UncompressedSize64)
+	}
+
+	var extractDir string
+	if sameLevel {
+		extractDir = filepath.Dir(zipPath)
+	} else {
+		extractDir = filepath.Join(filepath.Dir(zipPath), strings.TrimSuffix(filepath.Base(zipPath), ".zip"))
+	}
+
+	var done int64
+	startTime := time.Now()
+
+	for _, f := range reader.File {
+		if strings.Contains(f.Name, "..") {
+			return errors.New("potentially malicious zip item path")
+		}
+		outPath := filepath.Join(extractDir, f.Name)
+
+		// Make directory if current entry is a folder
+		if f.FileInfo().IsDir() {
+			if err := os.MkdirAll(outPath, f.Mode()); err != nil {
+				return err
+			}
+		}
+	}
+
+	for i, f := range reader.File {
+		if strings.Contains(f.Name, "..") {
+			return errors.New("potentially malicious zip item path")
+		}
+
+		// Already handled above
+		if f.FileInfo().IsDir() {
+			continue
+		}
+
+		outPath := filepath.Join(extractDir, f.Name)
+
+		// Otherwise create necessary parent directories
+		if err := os.MkdirAll(filepath.Dir(outPath), 0755); err != nil {
+			return err
+		}
+
+		// Open the file inside the archive
+		fileInArchive, err := f.Open()
+		if err != nil {
+			return err
+		}
+		defer fileInArchive.Close()
+
+		dstFile, err := os.Create(outPath)
+		if err != nil {
+			return err
+		}
+
+		// Read from zip in chunks to update progress
+		buffer := make([]byte, MiB)
+		for {
+			n, readErr := fileInArchive.Read(buffer)
+			if n > 0 {
+				_, writeErr := dstFile.Write(buffer[:n])
+				if writeErr != nil {
+					dstFile.Close()
+					return writeErr
```

Comment:
@Retengart 

Maybe we should add
```go
os.Remove(dstFile.Name())
```
here before returning to remove any incomplete files?